### DISCUSSION TITLE
Report Manager ✨

### DISCUSSION THREAD

**Discussion by abuzuhri** [abuzuhri/Amazon-SP-API-CSharp#76] [Announcements]
### Report Manager 🚀🧑‍🚀✨
Easy way to get the report you need and convert the file return from amazon to class or list, this feature only ready for some reports as its will take much times to finish for all reports ....
```CSharp
ReportManager reportManager = new ReportManager(amazonConnection);
var products = reportManager.GetProducts(); //GET_MERCHANT_LISTINGS_ALL_DATA
var inventoryAging = reportManager.GetInventoryAging(); //GET_FBA_INVENTORY_AGED_DATA
var ordersByDate = reportManager.GetOrdersByOrderDate(90); //GET_FLAT_FILE_ALL_ORDERS_DATA_BY_ORDER_DATE_GENERAL
var ordersByLastUpdate = reportManager.GetOrdersByLastUpdate(90); //GET_FLAT_FILE_ALL_ORDERS_DATA_BY_LAST_UPDATE_GENERAL
var settlementOrder = reportManager.GetSettlementOrder(90); //GET_V2_SETTLEMENT_REPORT_DATA_FLAT_FILE_V2
var returnMFNOrder = reportManager.GetReturnMFNOrder(90); //GET_FLAT_FILE_RETURNS_DATA_BY_RETURN_DATE
var returnFBAOrder = reportManager.GetReturnFBAOrder(90); //GET_FBA_FULFILLMENT_CUSTOMER_RETURNS_DATA
var reimbursementsOrder = reportManager.GetReimbursementsOrder(180); //GET_FBA_REIMBURSEMENTS_DATA
var feedbacks = reportManager.GetFeedbackFromDays(180); //GET_SELLER_FEEDBACK_DATA
```

I need more volunteer to test report and help to add more reports with same way 

Report Type List
https://developer-docs.amazon.com/sp-api/docs/report-type-values

**Comment by abuzuhri** [comment]
Hello @igoryan-k @J-W-Chan @kevinvenclovas @Kinggrass @skl-hyf @jsaxdev @Kinggrass @stokebob @GuybrushX @J-W-Chan @KristopherMackowiak 📣🚩


I'm asking for help to list all report to easy access and convert amazon file to class or list .

**Reply by asifiq2000** [reply]
Hello can I get (GET_FBA_MYI_ALL_INVENTORY_DATA) this report using GetReport() method else what is the way to get it, please help me out

**Reply by abuzuhri** [reply]
No that report you need to request separate, look for report sample file for how to get report

**Reply by asifiq2000** [reply]
Hello 

Query 1. As per given list this report (GET_MERCHANT_LISTINGS_ALL_DATA) is available in your plugin as mentioned below
var products = reportManager.GetProducts(); //GET_MERCHANT_LISTINGS_ALL_DATA
but I need to get this report (GET_MERCHANT_LISTINGS_DATA) so can I get this report too using the same method,

Query 2. I am getting below error 
"errors": [
    {
      "message": "Access to requested resource is denied.",
     "code": "Unauthorized",
     "details": ""
    }
  ]
while creating GET_FBA_MYI_ALL_INVENTORY_DATA report 
wherein I am passing 
restClient = new RestClient("https://sellingpartnerapi-na.amazon.com");
            IRestRequest restRequest = new RestRequest("/reports/2022-06-30/reports", Method.POST);
            restRequest.AddHeader("x-amz-access-token", str_token_from_your_secure_db);
            restRequest.AddHeader("reportType", "GET_V2_SETTLEMENT_REPORT_DATA_FLAT_FILE_V2");
            restRequest.AddHeader("marketplaceIds", "ATVPDKIKX0DER");
            var request = signatureHelper.SignRequest(restRequest, restClient, content_form_urlencoded);
            var result = restClient.Execute(request);
![image](https://user-images.githubusercontent.com/40463094/186271307-825c58e0-1552-4184-a8e4-5c19eda3bf98.png)

Also tried passing as parameters but getting different error
restRequest.AddQueryParameter("reportType", "GET_FBA_MYI_ALL_INVENTORY_DATA");
restRequest.AddQueryParameter("marketplaceIds", "ATVPDKIKX0DER");

Please help me how to pass these parameter to create a report and get this report from SP-API
or could you please provide any link where I get the sample request for the same. coz I did not 
get sample request in SP-API document.
Thanks in advance.

**Comment by skl-hyf** [comment]
![image](https://user-images.githubusercontent.com/76997845/186291387-9b3be4f4-e6a0-4f44-a824-bcd78cfaf75e.png)

As shown in the official documentation, the parameter should be placed in the Body of the POST request, and the url version of the request is /reports/2021-06-30/reports, you have used the wrong value 2022

**Reply by asifiq2000** [reply]
Thanks but what if I have to get the 30Jun2022 report then what needs to be passed into the request url /reports/2022-06-30/reports

Yes I saw the SP-API documentation mentioned to pass in body but I do not know how to frame a body for it, could you please help me out with a sample request body.

**Reply by skl-hyf** [reply]
You can try this.
https://github.com/abuzuhri/Amazon-SP-API-CSharp/blob/main/Source/FikaAmazonAPI/Parameter/Report/ParameterCreateReportSpecification.cs

**Reply by asifiq2000** [reply]
Hi 
Is there a way to make a request header and body to be posted using Postman 
I mean without using Fika Api source or plugin. 
Thanks for your answers, please help me out 
Actually Fika Api provides limited reports so for
other reports I need to create and get that report
using directly without Fika api, wherein I am facing 
issues, so please help me if I am able to make a request
for one report then I can do the rest accordingly.
please help with the sample request header and body just 
for creating report.

Awaiting your reply

**Reply by J-W-Chan** [reply]
> Hi Is there a way to make a request header and body to be posted using Postman I mean without using Fika Api source or plugin. Thanks for your answers, please help me out Actually Fika Api provides limited reports so for other reports I need to create and get that report using directly without Fika api, wherein I am facing issues, so please help me if I am able to make a request for one report then I can do the rest accordingly. please help with the sample request header and body just for creating report.
> 
> Awaiting your reply

you can see this  [link](https://developer-docs.amazon.com/sp-api/docs/using-postman-for-selling-partner-api-models)

**Comment by RenzoF** [comment]
Hi @abuzuhri I really like the Report Manager, though it might be good to be on its own package/repo and depends/consumes [Amazon-SP-API-CSharp](https://github.com/abuzuhri/Amazon-SP-API-CSharp)

**Reply by abuzuhri** [reply]
Good idea I will think on it

**Comment by kevinvenclovas** [comment]
Hi @abuzuhri , love the report manger but it works only with english report language configuration in the seller center.

![image](https://user-images.githubusercontent.com/7681603/226141679-b7bbf2ef-cc9e-43bb-b758-867bbd8b2b5a.png)

For my global Amazon App i tried to add new languages to make the ProductsReport more dynamic:
![image](https://user-images.githubusercontent.com/7681603/226141653-6888dd77-7b2c-469d-8da0-ba8e29f40d57.png)

After some languages i have a big question. **Why we use the table keys and not the table index?** In the ProductReport every language has the same arrangement of datas. Are there reports where it is not so or why do we use the table keys?

It would be much easier to write GetString(0) to become the item name :D

**Reply by abuzuhri** [reply]
Hello ,

At start we use index and it's was working good ,

After more work place test we found that report column change for market place to another at same report (some column missing), this why we move to use name to solve the problem.

**Reply by kevinvenclovas** [reply]
Oh oke so the best solution is to add all keys from all languages :/

**Reply by abuzuhri** [reply]
Yes key the best , if we return to index you will see date inside product name ....

For multi language we need to think In solution

**Comment by kevinvenclovas** [comment]
Oh i see some report used this kind of logic already :D 

![image](https://user-images.githubusercontent.com/7681603/226142428-2189dce3-9315-469d-9db7-89e158a6a70c.png)
